File: cs/reader.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using Mal;
using MalVal = Mal.types.MalVal;
using MalSymbol = Mal.types.MalSymbol;
using MalList = Mal.types.MalList;
using MalVector = Mal.types.MalVector;
using MalHashMap = Mal.types.MalHashMap;
using MalThrowable = Mal.types.MalThrowable;
using MalContinue = Mal.types.MalContinue;

namespace Mal {
    public class reader {
        public class ParseError : MalThrowable {
            public ParseError(string msg) : base(msg) { }
        }

        public class Reader {
            List<string> tokens;
            int position;
            public Reader(List<string> t) {
                tokens = t;
                position = 0;
            }

            public string peek() {
                if (position >= tokens.Count) {
                    return null;
                } else {
                    return tokens[position];
                }
            }
            public string next() {
                return tokens[position++];
            }
        }

        public static List<string> tokenize(string str) {
            List<string> tokens = new List<string>();
            string pattern = @"[\s ,]*(~@|[\[\]{}()'`~@]|""(?:[\\].|[^\\""])*""|;.*|[^\s \[\]{}()'""`~@,;]*)";
            Regex regex = new Regex(pattern);
            foreach (Match match in regex.Matches(str)) {
                string token = match.Groups[1].Value;
                if ((token != null) && !(token == "") && !(token[0] == ';')) {
                    //Console.WriteLine("match: ^" + match.Groups[1] + "$");
                    tokens.Add(token);
                }
            }
            return tokens;
        }

        public static MalVal read_atom(Reader rdr) {
            string token = rdr.next();
            string pattern = @"(^-?[0-9]+$)|(^-?[0-9][0-9.]*$)|(^nil$)|(^true$)|(^false$)|^("".*"")$|(^[^""]*$)";
            Regex regex = new Regex(pattern);
            Match match = regex.Match(token);
            //Console.WriteLine("token: ^" + token + "$");
            if (!match.Success) {
                throw new ParseError("unrecognized token '" + token + "'");
            }
            if (match.Groups[1].Value != String.Empty) {
                return new Mal.types.MalInteger(int.Parse(match.Groups[1].Value));
            } else if (match.Groups[3].Value != String.Empty) {
                return Mal.types.Nil;
            } else if (match.Groups[4].Value != String.Empty) {
                return Mal.types.True;
            } else if (match.Groups[5].Value != String.Empty) {
                return Mal.types.False;
            } else if (match.Groups[6].Value != String.Empty) {
                string str = match.Groups[6].Value;
                str = str.Substring(1, str.Length-2)
                    .Replace("\\\"", "\"")
                    .Replace("\\n", "\n");
                return new Mal.types.MalString(str);
            } else if (match.Groups[7].Value != String.Empty) {
                return new Mal.types.MalSymbol(match.Groups[7].Value);
            } else {
                throw new ParseError("unrecognized '" + match.Groups[0] + "'");
            }
        }

        public static MalVal read_list(Reader rdr, MalList lst, char start, char end) {
            string token = rdr.next();
            if (token[0] != start) {
                throw new ParseError("expected '" + start + "'");
            }

            while ((token = rdr.peek()) != null && token[0] != end) {
                lst.conj_BANG(read_form(rdr));
            }

            if (token == null) {
                throw new ParseError("expected '" + end + "', got EOF");
            }
            rdr.next();

            return lst;
        }

        public static MalVal read_hash_map(Reader rdr) {
            MalList lst = (MalList)read_list(rdr, new MalList(), '{', '}');
            return new MalHashMap(lst);
        }


        public static MalVal read_form(Reader rdr) {
            string token = rdr.peek();
            if (token == null) { throw new MalContinue(); }
            MalVal form = null;

            switch (token) {
                case "'": rdr.next();
                    return new MalList(new MalSymbol("quote"),
                                       read_form(rdr));
                case "`": rdr.next();
                    return new MalList(new MalSymbol("quasiquote"),
                                       read_form(rdr));
                case "~":
                    rdr.next();
                    return new MalList(new MalSymbol("unquote"),
                                       read_form(rdr));
                case "~@":
                    rdr.next();
                    return new MalList(new MalSymbol("splice-unquote"),
                                       read_form(rdr));
                case "^": rdr.next();
                    MalVal meta = read_form(rdr);
                    return new MalList(new MalSymbol("with-meta"),
                                       read_form(rdr),
                                       meta);
                case "@": rdr.next();
                    return new MalList(new MalSymbol("deref"),
                                       read_form(rdr));

                case "(": form = read_list(rdr, new MalList(), '(' , ')'); break;
                case ")": throw new ParseError("unexpected ')'");
                case "[": form = read_list(rdr, new MalVector(), '[' , ']'); break;
                case "]": throw new ParseError("unexpected ']'");
                case "{": form = read_hash_map(rdr); break;
                case "}": throw new ParseError("unexpected '}'");
                default:  form = read_atom(rdr); break;
            }
            return form;
        }


        public static MalVal read_str(string str) {
            return read_form(new Reader(tokenize(str)));
        }
    }
}

File: vb/reader.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Text.RegularExpressions
Imports Mal
Imports MalVal = Mal.types.MalVal
Imports MalSymbol = Mal.types.MalSymbol
Imports MalList = Mal.types.MalList
Imports MalVector = Mal.types.MalVector
Imports MalHashMap = Mal.types.MalHashMap
Imports MalThrowable = Mal.types.MalThrowable
Imports MalContinue = Mal.types.MalContinue

Namespace Mal
    Public Class reader
        Public Class ParseError
            Inherits MalThrowable
            Public Sub New(msg As String)
                MyBase.New(msg)
            End Sub
        End Class

        Public Class Reader
            Private tokens As New List(Of String)
            Private position As Int32 = 0
            Sub New(t As List(Of String))
                tokens = t
                position = 0
            End Sub

            Public Function peek() As String
                If position >= tokens.Count Then
                    return Nothing
                Else
                    return tokens(position)
                End If
            End Function

            Public Function get_next() As String
                If position >= tokens.Count Then
                    return Nothing
                Else
                    position += 1
                    return tokens(position-1)
                End If
            End Function
        End Class

        Shared Function tokenize(str As String) As List(Of String)
            Dim tokens As New List(Of String)
            Dim pattern As String = "[\s ,]*(~@|[\[\]{}()'`~@]|""(?:[\\].|[^\\""])*""|;.*|[^\s \[\]{}()'""`~@,;]*)"
            Dim regex As New Regex(pattern)
            For Each match As Match In regex.Matches(str)
                Dim token As String = match.Groups(1).Value
                If Not token Is Nothing _
                    AndAlso Not token = "" _
                    AndAlso Not token(0) = ";" Then
                    'Console.WriteLine("match: ^" & match.Groups[1] & "$")
                    tokens.Add(token)
                End If
            Next
            return tokens
        End Function

        Shared Function read_atom(rdr As Reader) As MalVal
            Dim token As String = rdr.get_next()
            Dim pattern As String = "(^-?[0-9]+$)|(^-?[0-9][0-9.]*$)|(^nil$)|(^true$)|(^false$)|^("".*"")$|^:(.*)|(^[^""]*$)"
            Dim regex As Regex = New Regex(pattern)
            Dim match As Match = regex.Match(token)
            'Console.WriteLine("token: ^" + token + "$")
            If not match.Success Then
                throw New ParseError("unrecognized token '" & token & "'")
            End If
            If match.Groups(1).Value <> String.Empty Then
                return New Mal.types.MalInt(Integer.Parse(match.Groups(1).Value))
            Else If match.Groups(3).Value <> String.Empty Then
                return Mal.types.Nil
            Else If match.Groups(4).Value <> String.Empty Then
                return Mal.types.MalTrue
            Else If match.Groups(5).Value <> String.Empty Then
                return Mal.types.MalFalse
            Else If match.Groups(6).Value <> String.Empty Then
                Dim str As String = match.Groups(6).Value
                return New Mal.types.MalString(
                        str.Substring(1, str.Length-2) _
                        .Replace("\""", """") _
                        .Replace("\n", Environment.NewLine) _
                        .Replace("\\", "\"))
            Else If match.Groups(7).Value <> String.Empty Then
                return New Mal.types.MalString(ChrW(&H029e) & match.Groups(7).Value)
            Else If match.Groups(8).Value <> String.Empty Then
                return New Mal.types.MalSymbol(match.Groups(8).Value)
            Else
                throw New ParseError("unrecognized '" & match.Groups(0).Value & "'")
            End If
        End Function

        Shared Function read_list(rdr As Reader, lst As MalList,
                                  start As String, last As String) As MalVal
            Dim token As String = rdr.get_next()
            If token(0) <> start Then
                throw New ParseError("expected '" & start & "'")
            End If
            
            token = rdr.peek()
            While token IsNot Nothing AndAlso token(0) <> last
                lst.conj_BANG(read_form(rdr))
                token = rdr.peek()
            End While

            If token Is Nothing Then
                throw New ParseError("expected '" & last & "', got EOF")
            End If
            rdr.get_next()

            return lst
        End Function

        Shared Function read_hash_map(rdr As Reader) As MalVal
            Dim lst As MalList = DirectCast(read_list(rdr, new MalList(),
                                                      "{", "}"),MalList)
            return New MalHashMap(lst)
        End Function


        Shared Function read_form(rdr As Reader) As MalVal
            Dim token As String = rdr.peek()
            If token Is Nothing Then
                throw New MalContinue()
            End If
            Dim form As MalVal = Nothing

            Select token
            Case "'"
                rdr.get_next()
                return New MalList(New MalSymbol("quote"),
                                   read_form(rdr))
            Case "`"
                rdr.get_next()
                return New MalList(New MalSymbol("quasiquote"),
                                   read_form(rdr))
            Case "~"
                rdr.get_next()
                return New MalList(New MalSymbol("unquote"),
                                   read_form(rdr))
            Case "~@"
                rdr.get_next()
                return new MalList(New MalSymbol("splice-unquote"),
                                   read_form(rdr))
            Case "^"
                rdr.get_next()
                Dim meta As MalVal = read_form(rdr)
                return new MalList(New MalSymbol("with-meta"),
                                   read_form(rdr),
                                   meta)
            Case "@"
                rdr.get_next()
                return new MalList(New MalSymbol("deref"),
                                   read_form(rdr))

            Case "("
                form = read_list(rdr, New MalList(), "(" , ")")
            Case ")"
                throw New ParseError("unexpected ')'")
            Case "["
                form = read_list(rdr, New MalVector(), "[" , "]")
            Case "]"
                throw New ParseError("unexpected ']'")
            Case "{"
                form = read_hash_map(rdr)
            Case "}"
                throw New ParseError("unexpected '}'")
            Case Else
                form = read_atom(rdr)
            End Select
            return form
        End Function


        Shared Function read_str(str As string) As MalVal
            return read_form(New Reader(tokenize(str)))
        End Function
    End Class
End Namespace
